File: Python/problem66.py
from math import sqrt

maxlim = 1000000;

squares = [];

def issquare(n):
	a = int(sqrt(n));
	return (a * a == n);

def minimalx(D):
	for i in xrange(2,maxlim):
		if (squares[i] == D):
			return 0;
		if ((squares[i] - 1) % D == 0):
			if (issquare((squares[i] - 1) / D)):
				return i;
	return -1;

def main():
	for i in xrange(maxlim):
		squares.append(i * i);
		if not issquare(squares[i]):
			print i
	largest = 0;
	Dlarge = 0;
	for i in xrange(2,1000+1):
		t = minimalx(i);
		if (t == -1):
			print "TOOLOW"
			#break;
		if (t > largest):
			largest = t;
			Dlarge = i;
	print Dlarge	

if __name__ == "__main__":
	main()

File: problem66.cpp
#include <iostream>
#include <math.h>
#include "bigint/BigIntegerLibrary.hh"

using namespace std;

const int maxlim = 20000;

BigInteger squares[maxlim];

bool issquare(BigInteger n) {
	int a = sqrt(n);
	return (a * a == n);
}

int minimalx(int D) {
	for (int i = 2; i < maxlim; i++) {
		if (squares[i] == D) {
			return 0;
		}
		if ((squares[i] - 1) % D == 0) {
			if (issquare((squares[i] - 1) / D)) {
				return i;
			}
		}
	}
	return -1;
}

int main() {
	for (int i = 0; i < maxlim; i++) {
		squares[i] = i * i;
		//cout << squares[i] << endl;
	}
	int largest = 0;
	int Dlarge = 0;
	for (int i = 2; i <= 1000; i++) {
		int t = minimalx(i);
		if (t == -1)
			cout << "TOO LOW";
		//cout << i << " " <<t << endl;
		if (t > largest) {
			largest = t;
			Dlarge = i;
			//cout << Dlarge << " " <<largest << endl;
		}
	}
	
	return 0;
}